[dist_autograd] expose distributed backward C++ API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38657 some refactor on register_distributed_ops
* #38494 [jit] register distributed backward
* **#38656 [dist_autograd] expose distributed backward C++ API**

Differential Revision: [D21940441](https://our.internmc.facebook.com/intern/diff/D21940441)

wanchaol committed:
[dist_autograd] expose distributed backward C++ API

[ghstack-poisoned]

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/autograd/autograd.cpp:
```diff
@@ -0,0 +1,24 @@
+
+} // namespace autograd
+} // namespace distributed
+} // namespace torch
```
Adding a new line after this line?

mrshenli commented on torch/csrc/distributed/autograd/autograd.cpp:
```diff
@@ -0,0 +1,24 @@
+
+void backward(
+    int64_t context_id,
+    const variable_list& roots,
```
This used to be a vector of Tensors instead of Variables, and then explicitly creating a vector of Variables. Will this cause any problem? Or are all call sites supposed to pass in Variables?

wanchaol replied:
I think all the callsites should pass in variables to begin with, although right now tensor/variable are the same thing.

mrshenli commented on torch/csrc/distributed/autograd/autograd.h:
```diff
@@ -1,11 +1,23 @@
+constexpr auto kDistAutogradBackwardProfilingKey =
+    "torch::distributed::autograd::backward";
+
+TORCH_API void backward(
```
Add some comment?

mrshenli reviewed (approved):
LGTM! Added some minor comments inline.

mrshenli reviewed (dismissed):
Build failures are real:

```
May 29 19:44:18 In file included from /var/lib/jenkins/workspace/torch/csrc/distributed/autograd/autograd.h:5:0,
May 29 19:44:18                  from /var/lib/jenkins/workspace/torch/csrc/distributed/autograd/autograd.cpp:1:
May 29 19:44:18 /var/lib/jenkins/workspace/torch/csrc/python_headers.h:10:10: fatal error: Python.h: No such file or directory
May 29 19:44:18  #include <Python.h>
```

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/autograd/autograd.h:
```diff
@@ -1,11 +1,38 @@
+    const variable_list& roots,
+    bool retain_graph = false);
+
 PyMethodDef* python_functions();
```
Should we split this file to separate python dependent part and pure C++ part?

wanchaol replied:
there were just a simple function for the python part so I didn't add it, I can split them for sure.

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

xush6528 reviewed (approved):
Saw tests passing this time.

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"




[ghstack-poisoned]

wanchaol committed:
Update on "[dist_autograd] expose distributed backward C++ API"


Differential Revision: [D21940441](https://our.internmc.facebook.com/intern/diff/D21940441)

[ghstack-poisoned]
